feat: 광고 감지 시 이동할 채널 선택 API 연동

devjaeyoon opened the pull request:
### ✨ 이슈 번호

- #128 

### 📌 설명

- 광고 감지 시 이동할 채널 선택 기능의 API를 연동한 Pull Request입니다.

### 📃 작업 사항

- [x] 광고 감지 시 이동할 채널 선택 API 연동
- [x] 광고 감지 시 이동할 채널 목록 정적 데이터에서 실제 채널 목록 중 isAdChannel값이 false인 채널로 리팩토링

### 💭 리뷰 사항 반영

- [x] 설정을 업데이트 할 때 모든 설정을 가져와 업데이트 하는 방식에서 업데이트하는 설정만 업데이트 하는 방식으로 변경

### ✅ Pull Request 체크 사항

- [x] 가장 최신 브랜치를 pull 하였습니다.
- [x] base 브랜치명을 확인하였습니다.
- [x] 코드 컨벤션을 모두 확인하였습니다.
- [x] 브랜치명을 확인하였습니다.

devjaeyoon committed:
feat: 광고 감지 시 이동할 채널 설정 기능 추가

- 광고 감지 시 이동할 채널(adRedirectChannelId) 설정 기능 UI 및 API 연동 구현
- 설정 관련 상수(SETTING_TYPES, SETTING_TITLES, SETTING_EXPLANATIONS) 확장
- useUserStore에 adRedirectChannelId 상태 추가 및 setUserSettings 수정
- useUpdateSetting 훅을 수정하여 value 전달 방식 지원
- Settings 페이지에서 광고 감지 여부에 따라 채널 선택 UI 조건부 렌더링

devjaeyoon committed:
refactor: 광고 감지 시 이동할 채널 목록을 동적 채널로 변경

기존에 하드코딩되어 있던 adRedirectChannelList를 제거하고,
useChannelStore에서 가져온  채널 목록 중
isAdChannel이 false인 채널만 필터링하여 사용하도록 수정함.
채널 변경 시 선택 로직 및 UI 동작은 기존 로직 그대로 유지됨.

devjaeyoon committed:
fix: 사용자 프로필에서 adRedirectChannelId 누락된 문제 수정

메인 화면 진입 시 사용자 정보를 불러올 때,
광고 감지 시 이동할 채널 ID(adRedirectChannelId)가 누락되어
설정이 제대로 반영되지 않는 문제가 있었음.
useUserProfile 훅에서 해당 필드를 포함하도록 수정.

devjaeyoon committed:
refactor: AD_REDIRECT_CHANNEL_ID를 AD_REDIRECT_CHANNEL로 변경

설정 중 하나인 광고 감지 시 이동할 채널은 상수명은 AD_REDIRECT_CHANNEL_ID보다
AD_REDIRECT_CHANNEL가 더 적합하기에 상수명을 수정함

d0-neii commented on src/hooks/useUpdateSetting.jsx:
```diff
@@ -7,19 +7,24 @@ import { useUserStore } from "@/store/useUserStore";
+
+    if (type === SETTING_TYPES.AD_REDIRECT_CHANNEL) {
+      updatedSettings[type] = value;
+    } else {
+      updatedSettings[type] = !settings[type];
+      if (
+        type === SETTING_TYPES.AD_DETECT &&
+        !updatedSettings[SETTING_TYPES.AD_DETECT]
+      ) {
+        updatedSettings[SETTING_TYPES.RETURN_CHANNEL] = false;
+      }
     }
```
`ad detect` 끄면 `return channel`도 같이 꺼지도록 처리해주신 부분 좋은 것 같아요!
그렇지만 처리 조건이 내부 if문으로 중첩된 구조라서,
`turnOffReturnChannelIfAdDetectOff(type, updatedSettings)`처럼 함수로 분리해두면
조건이 더 잘 드러나고 전체 흐름도 더 명확해질 것 같아요 :)

@faber-h, @devjaeyoon 두 분 의견도 궁금합니다~

faber-h replied:
중첩된 조건문을 함수로 분리하면 의도가 더 잘 드러나고 가독성도 좋아질 것 같아요.
다은님 말씀처럼 별도 함수로 추출하는 방향 좋은 것 같습니다!

다만 `turnOffReturnChannelIfAdDetectOff`는 함수명이 조금 길고 복잡해서,
가독성 측면에서는 약간 부담될 수 있을 것 같아요.

`adDetectToggle`처럼 더 간결하게 의도를 표현할 수 있는 이름도 같이 고민해보면 좋을 것 같습니다 😊

devjaeyoon replied:
해당 의견과 아래 의견을 종합하여 반영해봤습니다. [아래 코멘트](https://github.com/Radio-Premium/RadioPremium-FE/pull/129#discussion_r2146819503) 참고 부탁드립니다!

faber-h commented on src/store/useUserStore.js:
```diff
@@ -7,10 +7,10 @@ export const useUserStore = create(
 
-      setUserSettings: ({ isAdDetect, isReturnChannel }) =>
-        set({ settings: { isAdDetect, isReturnChannel } }),
+      setUserSettings: (updatedSettings) => set({ settings: updatedSettings }),
```
```suggestion
      setUserSettings: (updatedSettings) =>
        set((state) => ({
          settings: {
            ...state.settings,
            ...updatedSettings,
          },
        })),
```
기존에 `isAdDetect`, `isReturnChannel` 두 개의 설정만 관리하던 구조에서, 설정 항목이 세 개로 늘어나면서 `updatedSettings`로 통합하신 부분은 좋았습니다!

다만, 현재 구현에서는 `updatedSettings`에 항목이 누락될 경우 설정이 사라지는 문제가 생길 수 있어요.

이 부분은 기존 상태를 spread (...state.settings)로 병합하는 방식으로 처리하면, updatedSettings에 일부 항목만 전달해도 안전하게 업데이트할 수 있는데 해당 방식으로 수정하는 게 어떨까요?

devjaeyoon replied:
코드를 작성하며 한 설정을 업데이트 할 때, 다른 설정까지 같이 가져오고 있다는 불편함은 인지하고 있었는데 희주님의 리뷰를 반영하면 이 불편함을 해결할 수 있겠네요. 아래 리뷰와 함께 반영하겠습니다. 좋은 리뷰 감사드립니다!

faber-h commented on src/hooks/useUpdateSetting.jsx:
```diff
@@ -7,19 +7,24 @@ import { useUserStore } from "@/store/useUserStore";
-
-    if (type === SETTING_TYPES.AD_DETECT && !updatedSettings.isAdDetect) {
-      updatedSettings.isReturnChannel = false;
+    const updatedSettings = { ...settings };
```
```suggestion
    const updatedSettings = { };
```
앞서 [제안](https://github.com/Radio-Premium/RadioPremium-FE/pull/129#discussion_r2146791020)드린 zustand setUserSettings를 spread (...state.settings) 방식으로 변경하면
이 부분에서도 전체 settings를 spread 할 필요 없이 변경된 값만 담아도 되기 때문에
`const updatedSettings = {};`와 같이 간단히 초기화해도 될 것 같아요!

faber-h commented on src/hooks/useUpdateSetting.jsx:
```diff
@@ -7,19 +7,24 @@ import { useUserStore } from "@/store/useUserStore";
+    if (type === SETTING_TYPES.AD_REDIRECT_CHANNEL) {
+      updatedSettings[type] = value;
+    } else {
+      updatedSettings[type] = !settings[type];
+      if (
+        type === SETTING_TYPES.AD_DETECT &&
+        !updatedSettings[SETTING_TYPES.AD_DETECT]
+      ) {
+        updatedSettings[SETTING_TYPES.RETURN_CHANNEL] = false;
+      }
```
```suggestion
      const toggledValue = !settings[type];
      updatedSettings[type] = toggledValue;
      
      if (type === SETTING_TYPES.AD_DETECT && !toggledValue) {
        updatedSettings[SETTING_TYPES.RETURN_CHANNEL] = false;
      }
```
`updatedSettings[type] = !settings[type];` 이후 바로 `updatedSettings` 값을 조건문에서 사용하는 흐름이 잘 읽히지 않는 것 같아요!

if문의 조건을 `settings`를 기준으로 판단하는 건 어떨까요?
`!settings[type]`를 `toggledValue` 같은 변수로 분리하면 가독성이 더 좋아질 것 같아요!

두 분의 의견이 궁금합니다!
@Eun0713 @devjaeyoon

d0-neii replied:
`toggledValue`로 분리한 쪽이 흐름도 더 자연스럽고, 의도도 잘 나타내는 것 같아요.
개인적으로 희주님이 말씀하신 방향이 읽는 입장에서 이해하기에 조금 더 수월할 거 같다고 생각합니다.

devjaeyoon replied:
두 분 모두 좋은 의견 감사드립니다.

```suggestion
    const isToggleType = type !== SETTING_TYPES.AD_REDIRECT_CHANNEL;
    const toggledValue = isToggleType ? !settings[type] : value;

    const updatedSettings = {
      [type]: toggledValue,
    };

    if (type === SETTING_TYPES.AD_DETECT && !toggledValue) {
      updatedSettings[SETTING_TYPES.RETURN_CHANNEL] = false;
    }
```

희주님의 `useUserStore.js` 리뷰를 반영하고 두 분의 의견을 종합하여 이렇게 만들어봤는데 어떠신가요? @Eun0713 @faber-h

faber-h replied:
`isToggleType`이랑 `toggledValue`로 분리한 부분 정말 좋습니다!
고생하셨습니다!

d0-neii replied:
코드 흐름이 더 명확해진 것 같아요~
수고하셨습니다.

devjaeyoon committed:
refactor: 사용자 설정 업데이트 로직 개선 및 상태 병합 방식 변경

- useUpdateSetting 훅에서 설정 타입에 따라 토글 여부를 구분하여 업데이트 처리
  - AD_REDIRECT_CHANNEL은 값 그대로 사용
  - 그 외 설정은 toggle 방식으로 처리
  - AD_DETECT 해제 시 RETURN_CHANNEL도 자동으로 false 처리
- useUserStore의 setUserSettings가 전체 교체 방식에서 부분 병합 방식으로 변경
  - 기존 설정을 유지하며 일부만 업데이트되도록 개선

Eun0713 reviewed (approved):
LGTM!

faber-h reviewed (approved):
LGTM 🔥
